### PR TITLE
[WIP] Adding Distributed RL Model Example

### PR DESCRIPTION
**Pull request by mrshenli**
In this example, the RL model is distributed across one agent and
multiple observers. Each observer has a replicated submodel of
the RL model which all connect to the submodel on the agent.

During training, this example uses distributed autograd to set
gradients for all submodels. Then, it uses RPC calls to collect
gradients from all observers to the agent, sums those gradients,
applies the gradient to the local dummy model on agent, and then
broadcast the model parameters back to the observers to update
their models.

### CODE REVIEW

File: distributed/rpc/rl/distributed_model.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        super(ObserverPolicy, self).__init__()
+        self.affine1 = nn.Linear(4, 128)
+        self.dropout = nn.Dropout(p=0.6)
+        self.affine2 = nn.Linear(128, 2)
+
+    def forward(self, x):
+        x = self.affine1(x)
+        x = self.dropout(x)
+        x = F.relu(x)
+        return self.affine2(x)
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: each observer applies four layers in the forward pass

File: distributed/rpc/rl/distributed_model.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+class AgentPolicy(nn.Module):
+    def __init__(self):
+        super(AgentPolicy, self).__init__()
+
+        self.saved_log_probs = []
+        self.rewards = []
+
+    def forward(self, action_scores):
+        return F.softmax(action_scores, dim=1)
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: each agent only applies a softmax.

File: distributed/rpc/rl/distributed_model.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            if done:
+                break
+
+class Agent:
+    def __init__(self, world_size):
+        self.ob_rrefs = []
+        self.agent_rref = RRef(self)
+        self.rewards = {}
+        self.saved_log_probs = {}
+        self.ob_policy = ObserverPolicy()
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: The agent also creates a dummy ObserverPolicy so that it can use the same optimizer to update all model parameters. Note that this ObserverPolicy never participates in the forward or backward pass. It is only used to apply the summed gradients.

File: distributed/rpc/rl/distributed_model.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+        # collect gradients from observers
+        futs = []
+        # should be able to optimize this loop with a gather
+        for ob_rref in self.ob_rrefs:
+            futs.append(_async_remote_method(Observer.get_gradients, ob_rref, ctx_id))
+
+        grads = [fut.wait() for fut in futs]
+        grads = [*zip(*grads)]
+        grads = [sum(grad) for grad in grads]
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: the above few lines just sums the grads from all observers.

File: distributed/rpc/rl/distributed_model.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        # instead of putting it into a separate ctx
+        # set grads for observer model
+        ob_params = list(self.ob_policy.parameters())
+        for g, p in zip(grads, ob_params):
+            p.grad = g
+
+        # set grads for agent model
+        ctx_grads = dist_autograd.get_gradients(ctx_id)
+        for p in self.policy.parameters():
+            p.grad = ctx_grads[p]
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: Then, we set the grad field for all parameters. (note that the ob_policy is a dummy model, which is only useful for grad updates)

File: distributed/rpc/rl/distributed_model.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+        self.optimizer.step()
+
+        ob_params = [p.data.detach() for p in self.ob_policy.parameters()]
+        # should be able to optimize this loop with a scatter
+        futs = []
+        for ob_rref in self.ob_rrefs:
+            futs.append(_async_remote_method(Observer.update_model, ob_rref, ob_params))
+        for fut in futs:
+            fut.wait()
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: params on both the dummy model and the AgentPolicy are updated. Now broadcast the dummy model params to all observers to perform updates there.

File: distributed/rpc/rl/distributed_model.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: I believe the above few steps can be replaced by c10d gather/scatter, and wrapped into your own version of distributed optimizer. In that way, it will be more efficient and look better.

File: distributed/rpc/rl/distributed_model.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+class Observer:
+    def __init__(self):
+        self.id = rpc.get_worker_info().id
+        self.env = gym.make('CartPole-v1')
+        self.env.seed(args.seed)
+        self.policy = ObserverPolicy()
+
+    def get_gradients(self, ctx_id):
+        all_grads = dist_autograd.get_gradients(ctx_id)
```
Comment (mrshenli) [inline distributed/rpc/rl/distributed_model.py]: After the distributed backward pass, the grads for the `ObserverPolicy` model lives in `dist_autograd.get_gradients(ctx_id)` on the observer. Retrieve them and give them to the agent.
